Explicitly raise InvalidResponse if commands and responses counts don't match

byroot opened the pull request:
As discussed in https://github.com/redis/redis-rb/pull/509

The idea here is to fail loudly if such unlikely race condition happened.

@djanowski is this more acceptable to you?

djanowski commented on lib/redis/errors.rb:
```diff
@@ -14,6 +14,10 @@ def initialize(reply_type)
+  # Raised when the response from the server is unparsable
+  class InvalidResponse < BaseError
+  end
+
```
`ProtocolError`?

byroot replied:
I wanted to use ProtocolError initially, but it's constructor is customized. I can refactor it though.

djanowski commented on lib/redis/pipeline.rb:
```diff
@@ -58,6 +58,10 @@ def without_reconnect(&blk)
     def finish(replies, &blk)
+      unless replies.size == futures.size
+        raise InvalidResponse, "#{futures.size} commands were enqueued, but only #{replies.size} responses were recieved"
+      end
```
Typo.

djanowski commented on lib/redis/pipeline.rb:
```diff
@@ -71,7 +75,7 @@ def finish(replies, &blk)
     class Multi < self
       def finish(replies)
-        exec = replies.last
+        exec = replies.pop
```
If possible, let's not mutate an argument.

byroot replied:
I can copy the array instead if you think it's best.

byroot commented:
I fixed the typo, and refactored `ProtocolError` so I can reuse it.

I honestly think the `pop` is simpler and better than the copy. I mean I get your point about not mutating arguments, but in this case it's a method responsible for parsing a response. So I really don't expect any further usage of that object.

That being said I don't think super strongly about it. So I can totally do a `slice` instead. Just let me know.

byroot commented:
We no longer need this because pipelines are serialized eagerly.
